x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -0,0 +1,336 @@
+    private final AtomicBoolean changed = new AtomicBoolean();
+    private final AtomicBoolean aborted = new AtomicBoolean();
+
+    DataNodeRequestSender(TransportService transportService, Executor esqlExecutor, CancellableTask rootTask) {
+        this.transportService = transportService;
+        this.esqlExecutor = esqlExecutor;
+        this.rootTask = rootTask;
+    }
+
+    final void startComputeOnDataNodes(
+        String clusterAlias,
+        Set<String> concreteIndices,
+        OriginalIndices originalIndices,
+        QueryBuilder requestFilter,
+        Runnable runOnTaskFailure,
+        ActionListener<ComputeResponse> listener
+    ) {
+        final long startTimeInNanos = System.nanoTime();
+        searchShards(rootTask, clusterAlias, requestFilter, concreteIndices, originalIndices, ActionListener.wrap(targetShards -> {
+            try (var computeListener = new ComputeListener(transportService.getThreadPool(), runOnTaskFailure, listener.map(profiles -> {
+                TimeValue took = TimeValue.timeValueNanos(System.nanoTime() - startTimeInNanos);
+                return new ComputeResponse(
+                    profiles,
+                    took,
+                    targetShards.totalShards(),
+                    targetShards.totalShards(),
+                    targetShards.skippedShards(),
+                    0
+                );
+            }))) {
+                for (TargetShard shard : targetShards.shards.values()) {
+                    for (DiscoveryNode node : shard.remainingNodes) {
+                        nodePermits.putIfAbsent(node, new Semaphore(1));
+                    }
+                }
+                pendingShardIds.addAll(targetShards.shards.keySet());
+                trySendingRequestsForPendingShards(targetShards, computeListener);
+            }
+        }, listener::onFailure));
+    }
+
+    private void trySendingRequestsForPendingShards(TargetShards targetShards, ComputeListener computeListener) {
+        changed.set(true);
+        final ActionListener<Void> listener = computeListener.acquireAvoid();
+        try {
+            while (sendingLock.tryLock()) {
+                try {
+                    if (changed.compareAndSet(true, false) == false) {
+                        break;
+                    }
+                    for (ShardId shardId : pendingShardIds) {
+                        if (targetShards.getShard(shardId).remainingNodes.isEmpty()) {
+                            final Exception failure = shardFailures.get(shardId);
+                            if (failure == null) {
+                                shardFailures.put(shardId, new ShardNotFoundException(shardId, "no shard copies found {}", shardId));
+                            }
+                            aborted.set(true);
+                        }
+                    }
+                    if (aborted.get()) {
```

Comment:
We can take this branch if we get a shard failure on a shard that has sent data at the same time as we finish the loop above. But the shard failure may not be in the list yet. Do you think it'd be easier to think about `aborted` if we calculated it from the list of shard failures? Like if it were:
```
List<ShardFailure> shardFailures;
record ShardFailure(ShardId, Exception, boolean abort) {}

boolean aborted() {
  return shardFailures.stream.anyMatch(ShardFailure::abort);
}
```

I always get confused about state state being places in two places.

Reply:
I guess if we're calculating stuff from it we should be using `addAll` more than looping adds? I dunno.

Reply:
Maybe ignore me. I've never been that good at this sort of code.

Reply:
Good suggestion. I think this will also help with partial_results. I pushed https://github.com/elastic/elasticsearch/pull/120774/commits/5d000d33b9fb6ff8ede259641b34962ba0bef097.